### PR TITLE
[NOMERGEY] 606 workflow submission actor

### PR DESCRIPTION
**Pull request by dvoet**
PR to give intermediate feedback to The President

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+                              val executionServiceDAO: ExecutionServiceDAO,
+                              val batchSize: Int,
+                              val credential: Credential,
+                              val pollInterval: FiniteDuration) extends Actor with WorkflowSubmission with LazyLogging {
+
+  import context._
+
+  scheduleNextWorkflowQuery
+
+  override def receive = {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: all the functions called in the receive should return Futures. The Futures must be piped back to the actor. More comments below

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    case LookForWorkflows =>
+      getUnlaunchedWorkflowBatch() map {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: this mapping approach does not work because if the future fails the map internals don't run so nothing will happen and the failure goes nowhere... zombie actor

None should not be a possible outcome, make that a different message. Then this can simply be
getUnlaunchedWorkflowBatch pipeTo self

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        case Some(msg) => self ! msg
+        case None => scheduleNextWorkflowQuery
+      }
+    case SubmitWorkflowBatch(workflowIds) =>
+      self ! submitWorkflowBatch(workflowIds)
+
+    case Status.Failure(t) => throw t // an error happened in some future, let the supervisor handle it
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: you have not made a supervisor

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+  scheduleNextWorkflowQuery
+
+  override def receive = {
+    case LookForWorkflows =>
+      getUnlaunchedWorkflowBatch() map {
+        case Some(msg) => self ! msg
+        case None => scheduleNextWorkflowQuery
+      }
+    case SubmitWorkflowBatch(workflowIds) =>
+      self ! submitWorkflowBatch(workflowIds)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: submitWorkflowBatch needs to return a Future

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+      unpackedWfs <- loadWorkflows(wfRecs) //reify to real workflows
+      assert(unpackedWfs.size == wfRecs.size) //there should be the right number of them
+      assert(unpackedWfs.forall(_.isDefined)) //and they should all have been found
+      toSubmit = unpackedWfs.map(_.get) //un-option them too
+    } yield {
+      (execSvcWfOpts, toSubmit)
+    }
+
+    //now we have a future of workflow options and the workflows themselves...
+    submitThingsFuture map {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: this Future is just dropped on the floor

**Comment by dvoet** [comment]
@helgridly I think the overall approach is right but you have some Future handling problems that I commented upon

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala (hunk 3 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: correct, this is copypasta!

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala (hunk 5 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: yup -- not finished here yet

**Comment by helgridly** [comment]
Thanks for the feedback - this is useful stuff. I am surprised that my monster for-comprehensions pass muster but I'm not going to complain!

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        wsOpt <- dataAccess.workspaceQuery.findByIdQuery(subRec.head.workspaceId).result
+        workspace <- wsOpt
+        billingProject <- dataAccess.rawlsBillingProjectQuery.load(RawlsBillingProjectName(workspace.namespace))
+      } yield {
+        (userOpt, wsOpt.headOption, billingProject)
+      }
+    }
+  }
+  
+  def buildWorkflowOptions(submissionId: UUID): Future[ExecutionServiceWorkflowOptions] = {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: I am not sure why you had to rewrite this from the original. Is it just because you need to lookup the workspace and the user? Then add just that and leave the rest alone. I think you are making more work for yourself. Also note that I have had bad luck mixing different types of monads in a for comprehension, e.g. futures and options. IntelliJ says it is ok but the compiler disagrees.

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: Yeah - no WorkspaceContext or UserInfo any more means I have to jump through hoops to reconstruct them from the things I have. The workspace is used both for the bucket name and to look up the billing project, and I ended up with something weird like:

```
val workspaceFuture = blah
workspaceFuture.map { //to get the workspace name }
workspaceFuture.map { //again to get the billing project }
```

Mapping over the same future multiple times felt weird to me.

Jeff also pointed out the monad-mixing thing, so I suspect this might get nuked in favour of something else anyway. Not entirely sure what though.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+            case svv: SubmissionValidationValue if svv.value.isDefined =>
+              svv.inputName -> svv.value.get
+          }
+          MethodConfigResolver.propertiesToWdlInputs(methodProps.toMap)
+        }
+
+        //yield the things we're going to submit to Cromwell
+        (wdl, wfRecs.map(_.id), wfInputsBatch, wfOpts)
+      }
+    }
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala:190]: Probably pull this out into a separate function since it's a bit of a mess. `GetMeTheStuffToSubmitToCromwell`.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+        //save error messages into failures and flip them to Failed
+        val failures = results collect {
+          case (wfRecId, Right(failure: ExecutionServiceFailure)) =>
+            //TODO: batchify this saveMessages?
+            dataAccess.workflowQuery.saveMessages(Seq(AttributeString(failure.message)), wfRecId)
+            wfRecId
+        }
+        dataAccess.workflowQuery.batchUpdateStatus(failures, WorkflowStatuses.Failed)
+      } map { _ => LookForWorkflows }
+    } recover {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: Is this the right place for this recover? I was toying with putting them on each intermediate future (i.e. one on `workflowBatchFuture` and one on `cromwellSubmission` but then I got confused by whether it'd pass the failures along and recover multiple times for the same failure.

File: src/main/scala/org/broadinstitute/dsde/rawls/monitor/BootMonitors.scala [hunk 9]
```diff
@@ -53,4 +53,6 @@ object BootMonitors extends LazyLogging {
       case t: Throwable => logger.error("Error starting submission monitor", t)
     }
   }
+
+  //FIXME: Handle the power going out!
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/monitor/BootMonitors.scala:57]: Still need to do this.

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 10]
```diff
@@ -964,33 +964,27 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: SlickDataSo
...
-            val execStatusFuture = executionServiceDAO.submitWorkflows(wdl, workflowInputsBatch, workflowOptions, userInfo) recover {
-              case t: Exception => {
-                val error = "Unable to submit workflows when creating submission"
-                logger.error(error, t)
-                entityInputsBatch.map { _ =>
-                  Right(ExecutionServiceFailure("error", s"error communcating with exec service: ${t.getMessage}", None))
-                }
-              }
-            }
+            val execStatusFuture = executionServiceDAO.submitWorkflows(wdl, workflowInputsBatch, workflowOptions, userInfo)
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Doge is helping to nuke this.

**Comment by helgridly** [comment]
@dvoet Could you take another look at this? Still not done but would like another sanity check.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+
+  //submit the batch of workflows with the given ids
+  def submitWorkflowBatch(workflowIds: Seq[Long]): Future[WorkflowSubmissionMessage] = {
+
+    val workflowBatchFuture = dataSource.inTransaction { dataAccess =>
+      for {
+      //Load a bunch of things we'll need to reconstruct information:
+      //The list of workflows in this submission
+        wfRecs <- dataAccess.workflowQuery.findWorkflowByIds(workflowIds).result
+        submissionId = wfRecs.head.submissionId
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: Turns out that Scala doesn't like assignments in for comprehensions: the semantics [are weird and mindbendy](http://stackoverflow.com/questions/29433727/the-for-comprehension-with-assignment-loops-in-wrong-order).

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: Screws with the assert on the following line, too.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+          dataAccess.workflowQuery.batchUpdateStatus(wfsWithASingleSubmission.map(_.id), WorkflowStatuses.Launching)
+          wfsWithASingleSubmission
+        }
+      }
+    }
+
+    //if we find any, next step is to submit them. otherwise, look again.
+    unlaunchedWfOptF.map {
+      case workflowRecs if workflowRecs.nonEmpty => SubmitWorkflowBatch(workflowRecs.map(_.id))
+      case _ => LookForWorkflows
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala:108]: there should be a message to schedule next workflow (that uses the poll interval) and use it here. The way it stands, this will be a tight busy wait loop.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala (hunk 11 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: get rid of the asserts, I find they make it hard to follow

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+      for {
+      //Load a bunch of things we'll need to reconstruct information:
+      //The list of workflows in this submission
+        wfRecs <- dataAccess.workflowQuery.findWorkflowByIds(workflowIds).result
+        submissionId = wfRecs.head.submissionId
+        _ = assert(wfRecs.forall(_.submissionId == submissionId)) //sanity check they're all in the same submission
+        unpackedWfOpts <- DBIO.sequence(wfRecs.map(dataAccess.workflowQuery.loadWorkflow)) //reify to real workflows
+        _ = assert(unpackedWfOpts.size == wfRecs.size) //there should be the right number of them
+        _ = assert(unpackedWfOpts.forall(_.isDefined)) //and they should all have been found
+        unpackedWfs = for {wfOpt <- unpackedWfOpts; wf <- wfOpt} yield {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: wrap the unpackedWfOpts and unpackedWfs shenanigans in a function. And it could probably be 1 line

DBIO.sequence(wfRecs.map(dataAccess.workflowQuery.loadWorkflow.map(_.getOrElse(throw...))))

I suggest an exception because it seems bad if unpackedWfs and wfRecs have different sizes

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala (hunk 7 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala:190]: I would say you should pull each for comp line out including all those assignments, e.g.

```
submitterOpt <- dataAccess.rawlsUserQuery.load(RawlsUserRef(RawlsUserSubjectId(submissionRec.submitterId)))
submitter = submitterOpt.get
```

could be

```
def loadSubmitter(id: String) = {
  dataAccess.rawlsUserQuery.load(RawlsUserRef(RawlsUserSubjectId(Id))).map(_.get)
}
....
submitter <- loadSubmitter(submissionRec.submitterId)
```

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala (hunk 8 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: yes, 1 recover at the end

**Comment by dvoet** [comment]
I think this generally looks alright

**Comment by dvoet** [comment]
@helgridly note my commit

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala (hunk 12 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala:108]: Great catch.

**Comment by helgridly** [comment]
Thanks for the feedback. I will hammer on this some more tomorrow, rebase, write tests. Doubt it'll be done by EOD but hopefully Wednesday.

Closing this PR for now as it should be PR'd into the `FEAT_xxx_bbq` branch.

File: src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+
+    import ExecutionJsonSupport.ExecutionServiceWorkflowOptionsFormat
+    val cromwellSubmission = for {
+      (wdl: String, workflowRecIds: Seq[Long], wfInputsBatch: Seq[String], wfOpts: ExecutionServiceWorkflowOptions) <- workflowBatchFuture
+      workflowSubmitResult <- executionServiceDAO.submitWorkflows(wdl, wfInputsBatch, Option(wfOpts.toJson.toString), getUserInfo(credential))
+    } yield {
+      workflowRecIds.zip(workflowSubmitResult)
+    }
+
+    cromwellSubmission flatMap { results =>
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/jobexec/WorkflowSubmissionActor.scala]: in the block of code below you are not chaining your database calls. Each call to dataAccess.XXX does not actually do anything. You need to chain them together with map, flatMap, andThen. As it stands, the only thing that will happen is saving the failures.

**Comment by helgridly** [comment]
I'm closing this again because it should be merged into the FEAT branch, which is actually #418.
